dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java

```diff
@@ -277,65 +284,70 @@ private BooleanFormula encodeForwardProgress(Program program, ProgressModel prog
-        final Map<MemoryObject, BigInteger> memObj2Addr = computeStaticMemoryLayout(memory);
-
-        final var enc = new ArrayList<BooleanFormula>();
-        for (final MemoryObject memObj : memory.getObjects()) {
-            // For all objects, their 'final' value fetched here represents their constant value.
-            final Formula addressVariable = context.encodeFinalExpression(memObj);
-            final BigInteger addressInteger = memObj2Addr.get(memObj);
-
-            if (addressVariable instanceof BitvectorFormula bitvectorVariable) {
-                final BitvectorFormulaManager bvmgr = fmgr.getBitvectorFormulaManager();
-                final int length = bvmgr.getLength(bitvectorVariable);
-                enc.add(bvmgr.equal(bitvectorVariable, bvmgr.makeBitvector(length, addressInteger)));
+        return encodeMemoryLayout(context.getTask().getProgram().getMemory());
+    }
+
+    private BooleanFormula encodeMemoryLayout(Memory memory) {
+        final BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        final EncodingHelper helper = new EncodingHelper(context.getFormulaManager());
+        final List<BooleanFormula> enc = new ArrayList<>();
+
+        // TODO: We could sort the objects to generate better encoding:
+        //  "static -> dynamic with known size -> dynamic with unknown size"
+        //  For the former two we can then statically assign addresses as we do now and
+        //  only the latter needs dynamic encodings.
+        final List<MemoryObject> memoryObjects = ImmutableList.copyOf(memory.getObjects());
+        for (int i = 0; i < memoryObjects.size(); i++) {
+            final MemoryObject cur = memoryObjects.get(i);
+            final Formula addr = context.address(cur);
+            final Formula size = context.size(cur);
+            final Formula alignment;
+
+            // Encode size & compute alignment
+            if (cur.isStaticallyAllocated()) {
+                enc.add(helper.equals(size, context.encodeFinalExpression(cur.size())));
+                alignment = context.encodeFinalExpression(cur.alignment());
             } else {
-                assert addressVariable instanceof IntegerFormula;
-                final IntegerFormulaManager imgr = fmgr.getIntegerFormulaManager();
-                enc.add(imgr.equal((IntegerFormula) addressVariable, imgr.makeNumber(addressInteger)));
+                // Non-allocated objects get size 0
+                enc.add(helper.equals(size,
+                                bmgr.ifThenElse(context.execution(cur.getAllocationSite()),
+                                        context.encodeExpressionAt(cur.size(), cur.getAllocationSite()),
+                                        helper.value(BigInteger.ZERO, helper.typeOf(size)))
+                        )
+                );
+                // Non-allocated objects with variable alignment get alignment 1
```

Comment:
Why not zero? do we have any constraint that alignments must be positive?

EDIT: I guess this is because the first object address matches the alignment, right?

Reply:
Alignment of 0 would mean you have to divide by 0... `x % 0` is ill-defined.